internal/pkg/earpointer/state.go

```diff
@@ -338,54 +256,124 @@ func (state *State) mergeFieldMap(oldMap ReferenceFieldMap,
 
 // Helper to obtain the partition info for "aref" (which should be a partition
 // representative)
-func (state *State) getPartitionInfo(aref Reference) *PartitionInfo {
+func (state *State) partitionInfo(aref Reference) *PartitionInfo {
 	v, ok := state.partitions[aref]
 	if !ok {
 		log.Fatalf("Reference [%s] is not a partition rep", aref)
 	}
 	return &v
 }
 
-// Helper to find a partition representative. It calls onfailure callback
-// to delegate the action on lookup failure.
-func (state *State) lookupPartitionRep(
-	aref Reference,
-	onfailure func(abs Reference) Reference) Reference {
-	ref := aref
-	rep, ok := state.parents[ref]
-	if !ok {
-		res := onfailure(ref)
-		return res
+// ValueReferenceOrNil returns the reference holding the value *r of an address r.
+// Return nil if such a value reference doesn't exists.
+func (state *State) ValueReferenceOrNil(addr Reference) Reference {
+	// In the heap, r --> *r is implemented as r[PointToMarker] = *r.
+	fmap := state.PartitionFieldMap(addr)
+	if v, ok := fmap[getDirectPointToField()]; ok {
+		return v
 	}
+	return nil
+}
 
-	// If we found the partition rep, return right away.
-	if rep == ref {
-		return rep
+// PartitionState represents the state that has been fully constructed, after which
+// no more mutation operations will be performed. Its internal data structures are
+// optimized for lookups only.
+type PartitionState struct {
+	// Inherit "partitions" and "parents" and the APIs to access them.
+	State
```

Comment:
If no mutations should be performed, why expose functions like Unify which updates the internals. You could have an unexposed copy of State and then expose the functions you want to. It might be simpler to just copy the internals that you needs over during GetPartitionState().

Reply:
Now Partitions does not inherit any APIs from state, and Partitions provides only read-only APIs. The original state definition has been made private to the package.